endOfEpoch/metachain/trigger.go

```diff
@@ -53,53 +50,72 @@ func NewEndOfEpochTrigger(args *ArgsNewMetaEndOfEpochTrigger) (*trigger, error)
 	}
 
 	return &trigger{
-		rounder:                       args.Rounder,
 		roundsPerEpoch:                args.Settings.RoundsPerEpoch,
 		epochStartTime:                args.GenesisTime,
-		syncTimer:                     args.SyncTimer,
 		epoch:                         args.Epoch,
 		roundsBetweenForcedEndOfEpoch: args.Settings.MinRoundsBetweenEpochs,
+		rounder:                       args.Rounder,
 	}, nil
 }
 
 // IsEndOfEpoch return true if conditions are fullfilled for end of epoch
 func (t *trigger) IsEndOfEpoch() bool {
-	t.rounder.UpdateRound(t.epochStartTime, t.syncTimer.CurrentTime())
-	currRoundIndex := t.rounder.Index()
+	return t.isEndOfEpoch
+}
 
-	if currRoundIndex == 0 {
-		return true
+// ForceEndOfEpoch sets the conditions ofr end of epoch to true in case of edge cases
+func (t *trigger) ForceEndOfEpoch(round int64) error {
+	if t.currentRoundIndex > round {
+		return endOfEpoch.ErrSavedRoundIsHigherThanSaved
 	}
-
-	if currRoundIndex > t.roundsPerEpoch {
-		t.epoch += 1
-		t.epochStartTime = t.rounder.TimeStamp()
-		return true
+	if t.currentRoundIndex == round {
+		return endOfEpoch.ErrForceEndOfEpochCanBeCalledOnNewRound
+	}
+	if t.currentRoundIndex+1 != round {
+		return endOfEpoch.ErrNewRoundIsNotTheNextRound
 	}
 
-	return false
-}
-
-// ForceEndOfEpoch sets the conditions ofr end of epoch to true in case of edge cases
-func (t *trigger) ForceEndOfEpoch() error {
-	oldRoundIndex := t.rounder.Index()
-	t.rounder.UpdateRound(t.epochStartTime, t.syncTimer.CurrentTime())
-	currRoundIndex := t.rounder.Index()
+	t.currentRoundIndex = round
 
-	if currRoundIndex < t.roundsBetweenForcedEndOfEpoch {
+	if t.currentRoundIndex-t.epochStartRound < t.roundsBetweenForcedEndOfEpoch {
 		return endOfEpoch.ErrNotEnoughRoundsBetweenEpochs
 	}
 
-	if oldRoundIndex != currRoundIndex {
-		return endOfEpoch.ErrForceEndOfEpochCanBeCalledOnNewRound
-	}
-
 	t.epochStartTime = t.rounder.TimeStamp()
 	t.epoch += 1
+	t.epochStartRound = t.currentRoundIndex
+	t.isEndOfEpoch = true
 
 	return nil
 }
 
+// Update processes changes in the trigger
+func (t *trigger) Update(round int64) {
+	if t.currentRoundIndex > round {
+		return
+	}
+	if t.currentRoundIndex == round {
+		return
+	}
+	if t.currentRoundIndex+1 != round {
```

Comment:
you can remove lines 94-99 as this condition covers both of them

Reply:
done